plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -285,60 +529,334 @@ public boolean deleteNuageVspDevice(DeleteNuageVspDeviceCmd cmd) {
+                }
+            } else {
+                if (!Strings.isNullOrEmpty(dc.getInternalDns1())) {
+                    dnsServers.add(dc.getInternalDns1());
+                }
+                if (!Strings.isNullOrEmpty(dc.getInternalDns2())) {
+                    dnsServers.add(dc.getInternalDns2());
+                }
+            }
         }
+        return dnsServers;
+    }
 
-        public String getNuageVspEntity() {
-            return nuageVspEntity;
+    public List<String> getGatewaySystemIds() {
+        List<String> gatewayIds = null;
+        String gatewaySystemIds = String.valueOf(_configDao.getValue(NuageVspManager.NuageVspConfigGateway.key()));
+        if (!Strings.isNullOrEmpty(gatewaySystemIds)) {
+            StringTokenizer tokens = new StringTokenizer(gatewaySystemIds, ",");
+            gatewayIds = new ArrayList<String>(tokens.countTokens());
+            while (tokens.hasMoreTokens())
+            {
+                gatewayIds.add(tokens.nextToken());
+            }
         }
+        return gatewayIds;
+    }
 
-        @Override
-        public void run() {
-            nuageVspSync.syncWithNuageVsp(nuageVspEntity);
+    @Override
+    public boolean preStateTransitionEvent(Status oldState, Status.Event event, Status newState, Host host, boolean status, Object opaque) {
+        return true;
+    }
+
+    @Override
+    public boolean postStateTransitionEvent(StateMachine2.Transition<Status, Status.Event> transition, Host vo, boolean status, Object opaque) {
+        // Whenever a Nuage VSP Host comes up, check if all CS domains are present and check if the CMS ID is valid
+        if (transition.getToState() == Status.Up && vo instanceof HostVO) {
+            auditHost((HostVO) vo);
         }
+        return true;
+    }
+
+    @Override
+    public boolean configure(String name, Map<String, Object> params) throws ConfigurationException {
+        initMessageBusListeners();
+        initNuageVspResourceListeners();
+        initNuageNetworkOffering();
+        initNuageVspVpcOffering();
+        Status.getStateMachine().registerListener(this);
+        return true;
+    }
+
+    @DB
+    private void initMessageBusListeners() {
+        // Create corresponding enterprise and profile in VSP when creating a CS Domain
+        _messageBus.subscribe(DomainManager.MESSAGE_ADD_DOMAIN_EVENT, new MessageSubscriber() {
+            @Override
+            public void onPublishMessage(String senderAddress, String subject, Object args) {
+                Long domainId = (Long) args;
+                Domain domain = _domainDao.findById(domainId);
+
+                try {
+                    _domainDao.acquireInLockTable(domain.getId());
+
+                    List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listAll();
+                    for (NuageVspDeviceVO nuageVspDevice : nuageVspDevices) {
+                        HostVO host = findNuageVspHost(nuageVspDevice.getHostId());
```

Comment:
Yes, this is possible, but this means we would add a view to the DB, which would contain this join.